Add more details about code review and writing reviewable code (#105)
NUbots/NUbook: src/book/03-guides/04-general/01-git.mdx

---
section: Guides
chapter: General
title: Git and PR Workflow

> description: How to contribute to a NUbots project using Git and GitHub PRs..

Reviewer: ```suggestion
description: How to contribute to a NUbots project using Git and GitHub PRs.
```

### Code Review

> When making a PR you can select a number of reviewers to review your code, they will get a notification. Please choose people that you feel will be able to understand the problem your code is trying to solve.

Reviewer 2: This overlaps a lot with the previous paragraph. Probably should consolidate them into one paragraph, I think.
Two suggestions for this paragraph anyway :p

```suggestion
When making a PR you can select a number of reviewers to review your code, who each get a notification. Choose people who will be able to understand the problem your code is trying to solve.
```
Or
```suggestion
When making a PR you can select a number of reviewers to review your code, who each get a notification. Ideally, choose people that you feel will be able to understand the problem your code is trying to solve.
```

## Writing Reviewable Code

Writing reviewable code is a skill that can be developed with practice. Here are some general guidelines to make your PR easier to review:

- **Make it atomic**: A reviewable PR should only address one issue or add one feature.
- **Keep it small**: Ideally, a reviewable PR should not change much more than 400 lines of code, anything above this super-exponentially increases the time it takes to review, and makes it more likely that an error will be missed.
- **Name it well**: A reviewable PR should be named starting with a verb saying what you did, what issue was fixed, or what feature added. Prefer the present tense, e.g. "Update dependency X to version Y" instead of "Updated dependency X to version Y".

To keep PRs small and atomic you may have to split your PR into multiple sub PRs. To do this, make a new branch branch from the original PR and move over some code. Then make a separate PR targeting the original PR's branch. Make sure to mention the original PR in the new PR's description, using `#123`, where `123` is the sub PR's number. After completing the sub PR, merge it back into the original PR.

> ## Handling Review and Reviewee Etiquette

Reviewer: ```suggestion
## Handling Reviews and Reviewee Etiquette
```

Here are some tips for handling reviews as the author of a PR:

- When a review comment is left on your PR, you may respond with a reply on GitHub, or with a comment in the code. Prefer comments in the code where appropriate, so in the future others don't have to read through several PRs and can instead just read the code.
- When applying changes, you can go to the **Files Changed** tab and see the comments with the context of the code. Here you can also batch apply suggestions, so they only take one commit.
- After addressing a review comment, you can click **Resolve Conversation** to hide the conversation thread and reduce visual clutter. It is the reviewer's responsibility to check that your changes adequately addressed their concerns.
- Don't be alarmed when a reviewer makes over a dozen comments on your PR, just work through them systematically, and don't hesitate to ask for help or clarification if you need more details.
- Finally, when all the comments have been addressed from a review that requested changes, you can request another review by clicking the **Re-request review** button next to the reviewer's name.

> ## Work in Progress PRs

Reviewer: This section now looks a bit out of place at the bottom here. I'd move it to after **Formatting and Buildkite Checks**.

If you push your changes to GitHub and get a message that the build failed due to formatting issues, try running the relevant format commands, commit any changes, and push to GitHub again. This will trigger a new build, which will check the formatting again. You won't be able to merge the PR until the build checks pass.

## Reviewing Code

> To review code on GitHub open up a PR and open the Files Changed tab. This will give you the diff of the PR. You can click on the lines of code in the diff to add a comment and/or a suggestion to those lines. You can either leave a single comment or start a review. In a review all comments you make will have pending written on them to say that they have not been posted yet. Please make each comment non-ambiguous and clearly state what your concern is.

Reviewer 2: ```suggestion
To review code on GitHub open up a PR and open the Files Changed tab. This will give you the diff of the PR. You can click on the lines of code in the diff to add a comment and/or a suggestion to those lines. You can either leave a single comment or start a review. In a review all comments you make will have pending written on them to say that they have not been posted yet. Try to be as clear as possible when making comments. The contributor should be able to understand if the comment is asking for clarification, or asking for changes.
```

> To review code on GitHub open up a PR and open the Files Changed tab. This will give you the diff of the PR. You can click on the lines of code in the diff to add a comment and/or a suggestion to those lines. You can either leave a single comment or start a review. In a review all comments you make will have pending written on them to say that they have not been posted yet. Please make each comment non-ambiguous and clearly state what your concern is.

Reviewer: ```suggestion
To review code on GitHub, go to the relevant PR page and open the **Files Changed** tab. This will show the <abbr title="The difference between the current files and the changes in the PR">diff</abbr> of the PR.

You can add <abbr title="General remarks or questions">comments</comments> and/or <abbr title="Specific code changes that can be added automatically if accepted by the PR author">suggestions</abbr> to one or more lines in the diff. For a single line, move your mouse cursor over the line and click the plus button that appears next to the line number. For multiple lines, click and drag the plus button down, starting on the first line. A Markdown editor will pop up to write your comment or suggestion. To make a code suggestion, click the button that looks like a file icon with **±** inside it, and then make your changes to the code snippet that is inserted. Click the **Preview** tab to preview your changes. When done, you can either leave a single comment or start a review. In a review, all comments you make will be pending until you finish your review, when they will all be posted at once.

Make each comment non-ambiguous and clearly state what your concern is. Be kind and make sure to word your comments constructively. Prefer "we" over "you" when making suggestions for what to do.
```

> When reading the code, if there is anything that confuses you or think is incorrect make a comment with all your concerns. For example, if there is a nontrivial block of code without a comment explaining it you should ask for the author to provide such a comment. Another example is if there is a block of code that could do something that the autor may have not considered you should inform the author, with a comment, on what these other effects may be.

Reviewer 2: Adding a sentence on how to review NUbook/documentation PRs somewhere is probably worthwhile IMO
```suggestion
When reading the code, if there is anything that confuses you or anything you think is incorrect, make a comment with all your concerns. For example, if there is a non-trivial block of code without a comment explaining it you should ask for the author to provide such a comment. Another example is if there is a block of code that could do something that the author may have not considered. In such a case, you should inform the author with a comment, on what these other effects may be. 

When reviewing documentation PRs, you should look for factual inaccuracies, omissions of important information, awkward sentences, and typos.
```

> When finished reading the code and you have a bunch of pending comments for your review you can do one of three things. The first is to request changes, this blocks the code from being merged until you approve the branch, even if others have approved it. The second is to approve the PR, this allows it to be merged into master, don't do this unless you are confident that the code is correct. The third is to leave a comment, do this if you are not too concerened about the comments you have made.

Reviewer: ```suggestion
After reading the code and adding your pending comments, click **Finish your review** at the top right. This will show a popup with an input for overall comments, as well as the three actions you can take to finish your review:

- **Request changes**: this blocks the code from being merged until you subsequently approve the PR in another review, even if others have approved it.
- **Approve**: this allows the PR to be merged, so don't do this unless you are confident that the changes are correct.
- **Comment**: this adds your comments without blocking or approving the changes. Do this if you are not too concerned about the comments you have made.
```

> When finished reading the code and you have a bunch of pending comments for your review you can do one of three things. The first is to request changes, this blocks the code from being merged until you approve the branch, even if others have approved it. The second is to approve the PR, this allows it to be merged into master, don't do this unless you are confident that the code is correct. The third is to leave a comment, do this if you are not too concerened about the comments you have made.

Reviewer 2: ```suggestion
Once you have read the code and written your comments, you can finalise the review in three ways. The first is to request changes, which blocks the code from being merged until you approve the branch, even if others have approved it. Your review may be dismissed, but this is generally discouraged. The second is to approve the PR, which allows it to be merged into master. Don't approve PRs unless you are _reasonably_ confident that the code is correct. The third is to leave a comment. Select this option if you are not too concerned about the comments you have made, or if you're not sure when you'll be able to approve the PR once the changes are made.
```

> Reviewing code that you may not be fully confident with and leaving comments on helps both yourself learn and the author make more readable code.

Reviewer: ```suggestion
Reviewing code that you may not be fully confident with and leaving comments asking for clarification helps you learn and helps the author write more readable code, which is better for everyone.
```

> Reviewing code that you may not be fully confident with and leaving comments on helps both yourself learn and the author make more readable code.

Reviewer 2: ```suggestion
Reviewing code that you may not be fully confident with and leaving comments on helps both yourself learn and the author make more readable code. Asking questions and seeking clarification to understand the code is encouraged.
```

> ## Writing Reviewable Code and Reviewee Etiquette

Reviewer: ```suggestion
## Writing Reviewable Code
```

> Writting reviewable code is an art, here are some guidelines that you should stick to to make it easier for reviews:

Reviewer: ```suggestion
Writing reviewable code is a skill that can be developed with practice. Here are some general guidelines that you should stick to make your PR easier to review:
```

> Writting reviewable code is an art, here are some guidelines that you should stick to to make it easier for reviews:

Reviewer 2: ```suggestion
Writting reviewable code is an art. Here are some guidelines to make it easier for reviewers:
```

> - **Atomicity**: A reviewable PR should only address one issue or add one feature.

Reviewer: ```suggestion
- **Make it atomic**: A reviewable PR should only address one issue or add one feature.
```

> - **Length**: A reviewable PR should not change much more than 400 lines of code, anything above this super-exponentioally increases the time it takes to review.

Reviewer: ```suggestion
- **Keep it small**: Ideally, a reviewable PR should not change much more than 400 lines of code, anything above this super-exponentially increases the time it takes to review, and makes it more likely that an issue will be missed.
```

> - **Length**: A reviewable PR should not change much more than 400 lines of code, anything above this super-exponentioally increases the time it takes to review.

Reviewer 2: while that _might_ be true, non-maths students mightn't even know what it means :p 
```suggestion
- **Length**: A reviewable PR should not change much more than 400 lines of code, anything above this exponentially increases the time it takes to review.
```

> - **Name**: A reviewable PR should have a verb saying what you did the the issue fixed or the feature added.

Reviewer: ```suggestion
- **Name it well**: A reviewable PR should be named starting with a verb saying what you did, what issue was fixed, or what feature was added. Prefer the present tense, e.g. "Update dependency X to version Y" instead of "Updated dependency X to version Y".
```

> To keep PRs small and atomic you may have to split your PR into multiple sub PRs. To do this make a new branch for the sub PR and move over some code. Then make a seperate PR and mention it in the original PR with `#123` where `123` is the PR number. After chanining the sub PR merge it back into the original PR.

Reviewer: ```suggestion
To keep PRs small and atomic you may have to split your PR into multiple sub PRs. To do this, make a new branch branch from the original PR and move over some code. Then make a separate PR targeting the original PR's branch. Make sure to mention the original PR in the new PR's description, using `#123`, where `123` is the original PR number. After completing the sub PR, merge it back into the original PR.
```

> To keep PRs small and atomic you may have to split your PR into multiple sub PRs. To do this make a new branch for the sub PR and move over some code. Then make a seperate PR and mention it in the original PR with `#123` where `123` is the PR number. After chanining the sub PR merge it back into the original PR.

Reviewer 2: is "chaining" the word you were looking for? I think this paragraph could use more detail, regardless
```suggestion
To keep PRs small and atomic you may have to split your PR into multiple sub PRs. To do this make a new branch for the sub PR and move over some code. Then make a seperate PR and mention it in the original PR with `#123` where `123` is the PR number. After chaining the sub PR merge it back into the original PR.
```

> 

Reviewer: ```suggestion

## Handling Reviews and Reviewee Etiquette

```

> When your code is being reviewed and reviewers add comments, if you can answer these concerns in the code with comments, do so, as in the furture people don't have to read through hundreds of PRs and can instead just read the code. When this is done you can press Resolve Conversation and this will hide the conversation. Don't be alarmed when a reviewer make over a dozen comments on your PR, just work through them systematically. Finally when all the comments have been addressed from a reviewer that has requested changes you can request that they review your code by pressing the Re-request review button.

Reviewer: ```suggestion
Here are some tips for handling reviews as the author of a PR:

- When a review comment is left on your PR, you may respond with a reply on GitHub, or with a comment in the code. Prefer comments in the code where appropriate, so in the future others don't have to read through several PRs and can instead just read the code.
- After addressing a review comment, you can click **Resolve Conversation** to hide the conversation thread and reduce visual clutter.
- Don't be alarmed when a reviewer makes over a dozen comments on your PR, just work through them systematically, and don't hesitate to ask for help or clarification if you need more details.
- Finally, when all the comments have been addressed from a review that requested changes, you can request another review by clicking the **Re-request review** button next to the reviewer's name.
```

> When your code is being reviewed and reviewers add comments, if you can answer these concerns in the code with comments, do so, as in the furture people don't have to read through hundreds of PRs and can instead just read the code. When this is done you can press Resolve Conversation and this will hide the conversation. Don't be alarmed when a reviewer make over a dozen comments on your PR, just work through them systematically. Finally when all the comments have been addressed from a reviewer that has requested changes you can request that they review your code by pressing the Re-request review button.

Reviewer 2: ```suggestion
When your code is being reviewed and reviewers add comments, if you can answer these concerns in the code with comments, do so, as in the future people don't have to read through hundreds of PRs and can instead just read the code. When this is done you can press `Resolve Conversation`, which hides the conversation. It is the reviewer's responsibility to check that your changes adequately addressed their concerns. 

Don't be alarmed when a reviewer make over a dozen comments on your PR. Work through them systematically. If you need any clarification on a review, reply to the comment asking for it. Replying in the thread allows other reviewers and future team members to see why a decision was made, although you should use your discretion as to when you should move conversations to a private forum. Finally, once all of a specific reviewer's comments have been addressed, you can request that they re-review your code by pressing the Re-request review button.
```

- **Name it well**: A reviewable PR should be named starting with a verb saying what you did, what issue was fixed, or what feature added. Prefer the present tense, e.g. "Update dependency X to version Y" instead of "Updated dependency X to version Y".

To keep PRs small and atomic you may have to split your PR into multiple sub PRs. To do this, make a new branch branch from the original PR and move over some code. Then make a separate PR targeting the original PR's branch. Make sure to mention the original PR in the new PR's description, using `#123`, where `123` is the sub PR's number. After completing the sub PR, merge it back into the original PR.

## Handling Reviews and Reviewee Etiquette

Here are some tips for handling reviews as the author of a PR:

- When a review comment is left on your PR, you may respond with a reply on GitHub, or with a comment in the code. Prefer comments in the code where appropriate, so in the future others don't have to read through several PRs and can instead just read the code.

> - When applying changes, you can go to the **Files Changed** tab and see the comments with the context of the code. Here you can also batch apply suggestions, so they only take one commit.

Reviewer: ```suggestion
- When applying suggested changes, you can go to the **Files Changed** tab and see the comments with the context of the code. Here you can also batch apply suggestions, so they only take one commit.
```
